Updating pbs_cgroups_hook.py to multi-node setup and added few more regressions

anamikau opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
updated pbs_cgroups_hook.py tests with multi-node setup and added few more regressions that were missing from the initial coverage like verifying that cgroups files are cleaned up, cgroups file created for multi-node jobs, etc ...

#### Affected Platform(s) and PBS Version
* *Linux*
* *PBS 18.1.1*

#### Cause / Analysis / Design
* *cgroup verification missing for multi node. Also few regression cases were not covered*

#### Solution Description
* *updated pbs_cgroups_hook.py tests with multi-node setup and added few more regressions that were missing from the initial coverage like verifying that cgroups files are cleaned up, cgroups file created for multi-node jobs, etc ...*

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

mike0042 commented:
Here's what the Travis CI logs are showing...

> test/tests/functional/pbs_cgroups_hook.py:572:25: E126 continuation line over-indented for hanging indent
>                         hostname=mom,
>                         ^
> test/tests/functional/pbs_cgroups_hook.py:573:25: E126 continuation line over-indented for hanging indent
>                         log=filename, sudo=True) as fd:
>                         ^

Codacy is complaining about unused variable 'ret1'.

latha-subramanian commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -1047,6 +1192,129 @@ def test_cgroup_reserve_mem(self):
+        rv1 = self.du.isdir(hostname=self.momA, path=cpath, sudo=True)
+        self.assertEqual(rv1, True)
+        rv2 = self.du.isdir(hostname=self.momB, path=cpath, sudo=True)
+        self.assertEqual(rv2, True)
```
would assertTrue() or assertFalse() be enough here rather than assertEqual ()

anamikau replied:
May be I am not sure. Any specific reason against using assertEqual?

latha-subramanian replied:
Just a convention, that I see,  used in other places of the file

bhroam replied:
While they are technically the same, I see the purpose of assertTrue() is to indeed assert something is true.  I think it is the right call to use here.

anamikau commented:
[ptl_cgroups_qa2.txt](https://github.com/PBSPro/pbspro/files/1692559/ptl_cgroups_qa2.txt)
Thanks @latha-subramanian and @mike0042 . addressed review comments.

anamikau commented:
Fixed codacy and some more. Attaching latest test execution results.
[ptl_cgroups_qa4.txt](https://github.com/PBSPro/pbspro/files/1693992/ptl_cgroups_qa4.txt)

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -440,6 +464,18 @@ def get_paths(self):
+        Returns True if path exists otherwise false
+        """
+        x = 0
+        while (x < 10):
```
You should use...

for x in range(10):

Since you don't actually use the value of x, you can also do the following:

for _ in range(10):

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -38,13 +38,36 @@
 from tests.functional import *
 
 
+def get_swap():
```
Would "have_swap" be a better name? It implies a boolean is returned.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -857,87 +972,118 @@ def test_cgroup_offline_node(self):
         """
         Test to verify that cgroups subsystems are not enforced on nodes
-        that have the exclude_hosts set
+        that have the exclude_hosts set but on other systems
```
"but are enforced on other systems"

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -857,87 +972,118 @@ def test_cgroup_offline_node(self):
+        jid2 = self.server.submit(j)
+        a = {'job_state': 'R'}
+        self.server.expect(JOB, a, jid2)
+        cpath = self.paths['cpuset'] + "/pbspro/" + jid2
```
Please use os.path.join()
There are several instances of this.

anamikau commented:
@mike0042 thanks for your comments. addressed them all. 
[ptl_cgroups_qa5.txt](https://github.com/PBSPro/pbspro/files/1698187/ptl_cgroups_qa5.txt)

latha-subramanian commented:
@anamikau , The changes look good to me. Thanks!

anamikau commented:
Thanks @mike0042 and @latha-subramanian . While running my tests on CRAY and HPE ICE I am noticing some failures hence I will leave the PR open and will not squash.

anamikau commented:
rebased and squashed. please take this forward once all checks will pass.
[ptl_cgroup1.txt](https://github.com/PBSPro/pbspro/files/1729788/ptl_cgroup1.txt)

anamikau commented:
Forgot to mention that I will address ICE and CRAY issue in a separate PR. There is no point of keeping this open.

mike0042 commented:
Restarted Debian build that encountered transient failure. Please rebase for merge.

anamikau commented:
something got merged hence rebased again

anamikau commented:
all checks have passed. @mike0042 pr @PBSPro/pbspro-maintainers can you take this forward.
